---
title: "cegwas results"
output: html_document
---



```{r setup, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)
library(cegwas)
library(dplyr)
library(ggplot2)
library(jsonlite)
library(RMySQL)

# Recieve Trait to pull
if (length(commandArgs(trailingOnly=TRUE)) != 0) {
  args = fromJSON(commandArgs(trailingOnly=TRUE))
} else {
  args = fromJSON('{"report_info": {"release": 0, "version": 0.1, "report_slug": "yyy", "email": "test", "report_name": "yyy"}, "trait_name": "trait1"}')
}
mysql_credentials <- fromJSON(readLines("db_credentials.json"))

# To connect to a database first create a src:
db <- src_mysql(dbname = "cegwas", host = mysql_credentials$host, user = mysql_credentials$user, password= mysql_credentials$password)
# Then reference a tbl within that src
trait <- collect(tbl(db, "report_trait_strain") %>%
             dplyr::select(strain, report_slug, name, value) %>%
             dplyr::filter(name == args$trait_name, report_slug == args$report_info$report_slug) %>%
             dplyr::select(strain, value))
names(trait) <- c("strain", args$trait_name)

dir.create(args$report_info$report_slug)

output <- opts_knit$get("rmarkdown.pandoc.to")
opts_chunk$set(warning=F, message=F, echo=F, results="hide")

output <- "html"
if (output=="html") opts_chunk$set(fig.width=12, fig.height=6)
```

#### Summary of everything to follow... . .. 

```{r summary}
trait <- process_pheno(trait)
mapping <- gwas_mappings(trait, mapping_snp_set = FALSE)
proc_mappings <- process_mappings(mapping,trait,BF = 5) %>% 
  dplyr::filter(log10p !=0) %>% 
  filter(!(grepl("MtDNA",marker) & log10p < BF))
```

## Distribution of Phenotype

```{r Phenotype Histogram}
ggplot2::ggplot(tidyr::gather(data.frame(trait[[2]]), strain, value))+
  ggplot2::aes(x = value)+
  ggplot2::geom_histogram(color = "black", fill = "grey60")+
  ggplot2::theme_bw() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(size=24, face="bold", color="black"),
                 axis.text.y = ggplot2::element_text(size=24, face="bold", color="black"),
                 axis.title.x = ggplot2::element_text(size=24, face="bold", color="black", vjust=-.3),
                 axis.title.y = ggplot2::element_text(size=24, face="bold", color="black"),
                 strip.text.x = ggplot2::element_text(size=24, face="bold", color="black"),
                 strip.text.y = ggplot2::element_text(size=16, face="bold", color="black"),
                 plot.title = ggplot2::element_text(size=24, face="bold", vjust = 1),
                 legend.position="none",
                 panel.background = ggplot2::element_rect( color="black",size=1.2),
                 strip.background = ggplot2::element_rect(color = "black", size = 1.2)) +
  ggplot2::labs(x = "Phenotype", y = "Count")
```

```{r}
if(nrow(proc_mappings) == 0){
  proc_mappings <- process_mappings(mapping,trait,BF = 5) %>% 
    dplyr::filter(log10p !=0) %>% 
    filter(!(grepl("MtDNA",marker) & log10p < BF))
  if(nrow(proc_mappings) == 0){
    proc_mappings <- process_mappings(mapping,trait,BF = 4.5) %>% 
      dplyr::filter(log10p !=0) %>% 
      filter(!(grepl("MtDNA",marker) & log10p < BF))
    if(nrow(proc_mappings) == 0){
      proc_mappings <- process_mappings(mapping,trait,BF = 4) %>% 
        dplyr::filter(log10p !=0) %>% 
        filter(!(grepl("MtDNA",marker) & log10p < BF))
      if(nrow(proc_mappings) == 0){
        proc_mappings <- process_mappings(mapping,trait,BF = 3.5) %>% 
          dplyr::filter(log10p !=0) %>% 
          filter(!(grepl("MtDNA",marker) & log10p < BF))
        if(nrow(proc_mappings) == 0){
          proc_mappings <- process_mappings(mapping,trait,BF = 3) %>% 
            dplyr::filter(log10p !=0) %>% 
            filter(!(grepl("MtDNA",marker) & log10p < BF))
        } else {
          proc_mappings <- as.data.frame()
        }
      }
    }
  }
}

```

`r if (nrow(proc_mappings) == 0) '# No Significant Mappings Identified'`

```{r}
if (nrow(proc_mappings) != 0){
  peaks <- na.omit(proc_mappings) %>%
  dplyr::distinct(peak_id) %>%
  dplyr::select(marker, CHROM, startPOS, endPOS, log10p, trait) %>%
  dplyr::mutate(query = paste0(CHROM, ":",startPOS, "-",endPOS)) %>%
  dplyr::arrange(desc(log10p)) %>%
  dplyr::mutate(top3peaks = seq(1:n())) %>%
  dplyr::filter(top3peaks < 4) %>%
  dplyr::select(trait,peak_pos = marker, interval = query, peak_log10p = log10p)
}
```

```{r}
if (nrow(proc_mappings) != 0){
  if(nrow(peaks) > 0){
    eval_cell_1 = TRUE
  } else{
    eval_cell_1 = FALSE
  }
  
  if(nrow(peaks) > 1){
    eval_cell_2 = TRUE
  } else{
    eval_cell_2 = FALSE
  }
  
  if(nrow(peaks) > 2){
    eval_cell_3 = TRUE
  } else{
    eval_cell_3 = FALSE
  }
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# Manhattan Plot'`
`r if (eval_cell_1) 'The figure below shows identified quantitative trait loci (QTL). Each dot represents a single-nucleotide variant (SNV) present in the _C.elegans_ population. Blue dots represent SNVs that are above the defined significance threshold, which is the thick gray line. Red boxes surrounding blue dots represent the QTL confidence interval, which we define as plus/minus 50 SNVs from the last signifincant SNV. The default threshold for significance is the Bonferroni-corrected value (-log10(.05/#SNVs)) and is usually around 5.5.'`
```{r Manplot}
if (nrow(proc_mappings) != 0){
  mplot <- manplot(proc_mappings)
  mplot[[1]]
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# QQ-plot'`
`r if (eval_cell_1) 'The figure below shows the signifcance values of all SNVs compared to a normally distributed data set. Deviations from this _y=x_ indicate that your data are not normally distributed.'`
```{r}
if (nrow(proc_mappings) != 0){
  log10ps <- distinct(proc_mappings, marker)
  
  cegwas::qq_plot(log10ps$log10p) + scale_x_continuous(limits = c(0, NA))
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# PxG plot'`
`r if (eval_cell_1) 'Below are boxplots showing the how your phenotype data are distributed when split at the peak SNV for each identified QTL. REF refers to the N2 genotype and ALT refers to the variant phenotype. All of the SNVs in our data set are biallelic so these are the only two classes for any given SNV. The phenotype is on the y-axis and the genotype is on the x-axis. QTL chromosome and position is indicated above each plot.'`
```{r PxGplot,echo=FALSE,  warning=FALSE, message=FALSE}
if (nrow(proc_mappings) != 0){
  pg_plot <- pxg_plot(proc_mappings, color_strains = NA) 
  pg_plot[[1]] +theme(legend.position = "none")
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# Tajimas for Most Significant QTL'`
`r if (eval_cell_1) 'Below is a plot showing calculated Tajimas D across the most signficant QTL interval. Tajimas D is tests whether a region of the genome has the expected level of variation within the population with values close to 0 indicating observed variation similar to expected levels, D<0 rare alleles are present that may have arose due to a recent selective sweep, population expansion after a recent bottleneck, or linkage to a swept gene, and D>0 indicating balancing selection or sudden population contraction.'`
```{r TajimasD,echo=FALSE,  warning=FALSE, message=FALSE, results="hide"}
if(nrow(peaks) > 0){
  for_tD <- tidyr::separate(peaks, interval,into = c("CHROM", "interval"), sep = ":") %>%
    tidyr::separate(interval, into = c("start", "end"), sep = "-") %>%
    tidyr::separate(peak_pos, into = c("ch", "peak_pos"), sep = "_")
  
  cegwas::tajimas_d(chromosome = for_tD$CHROM[1],
            interval_start = for_tD$start[1],
            interval_end = for_tD$end[1],
            site_of_interest = as.numeric(for_tD$peak_pos[1]))
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# Global Distribution of Peak Variant'`
`r if (eval_cell_1) 'Below is the world-wide distribution of the SNV most highly correlated with your phenotype.'`
```{r Global Distribution,echo=FALSE,  warning=FALSE, message=FALSE, results="hide"}
if(nrow(peaks) > 0){
  for_map <- gsub(pattern = "_", replacement = ":", peaks$peak_pos)
  cegwas::allele_distribution(for_map[1])
}
```

`r if (eval_cell_2) '------------------------'`

`r if (eval_cell_2) '# Linkage Disequilibrium Between Peak Markers'`
`r if (eval_cell_2) 'Below is a plot representing the linkage disequilibrium (LD) between QTL. The LD measure being plotted is the correlation between markers _r = -D / sqrt( p(A) * p(a) * p(B) * p(b) )_, where D raw difference in frequency between the observed number of AB pairs and the expected number and A, B, a, and b refer to alleles at the loci. As a rule of thumb - an LD value of .8 or higher is suggests that the two peak markers are not segregating randomly.'`
```{r LDplot,echo=FALSE,  warning=FALSE, message=FALSE, results="hide"}
if(nrow(peaks) > 1){
plot_peak_ld(proc_mappings)
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '# Summary of QTL Confidence Intervals'`

`r if (eval_cell_1) 'Below are summary tables of the identified QTL. The first table is a summary of identified QTL and the corresponding confidence intervals. The following tables contain gene variants that are highly correlated with your phenotype present in each of the identified QTL (only the top three QTL are summarised for clarity). The tables are generated by scanning the QTL confidence interval for variants present in the _C. elegans_ population. Next, the Spearman rank correlation coefficient is calculated to test each variants correlation with your phenotype of interest. The variants with the strongest correlation are reported in the tables.'`

```{r peaks, echo = FALSE, warning=FALSE, message=FALSE}
if(nrow(peaks) > 0){
    DT::datatable(peaks,
                extensions = "Responsive")
} else {
  print("NO QTL IDENTIFIED")
}

```

```{r, cache=T, cache.path="cache/", warning=FALSE, message=FALSE, results="hide"}
if(nrow(proc_mappings) != 0){
interval_variants <- process_correlations(variant_correlation(proc_mappings, quantile_cutoff_high = 0.75, quantile_cutoff_low = 0.25))
}
```

`r if (eval_cell_1) '------------------------'`

`r if (eval_cell_1) '## Peak 1'`

```{r Peak1, warning=FALSE, message=FALSE}
if(nrow(proc_mappings) != 0){
  peak_1 <- interval_variants %>%  
    dplyr::mutate(query = paste0(CHROM, ":",startPOS, "-",endPOS)) %>% 
    dplyr::filter(query == peaks$interval[1]) %>%
    dplyr::select(CHROM, POS, INTERVAL = query, GENE_NAME = gene_name, REF, ALT, 
                  NT_CHANGE = nt_change, AA_CHANGE = aa_change, GENE_ID = gene_id, 
                  TRANSCRIPT = molecular_name, DESCRIPTION = concise_description, 
                  CORRELATION = abs_spearman_cor) %>%
    dplyr::distinct()
  
  DT::datatable(peak_1,
                filter = "top",
                extensions = "Responsive")
}

```

`r if (eval_cell_2) '------------------------'`

`r if (eval_cell_2) '## Peak 2'`

```{r Peak2,echo=FALSE, warning=FALSE, message=FALSE}
if(nrow(peaks) > 1){
  peak_2 <- interval_variants %>%  
  dplyr::mutate(query = paste0(CHROM, ":",startPOS, "-",endPOS)) %>% 
  dplyr::filter(query == peaks$interval[2]) %>%
    dplyr::select(CHROM, POS, INTERVAL = query, GENE_NAME = gene_name, REF, ALT, 
                  NT_CHANGE = nt_change, AA_CHANGE = aa_change, GENE_ID = gene_id, 
                  TRANSCRIPT = molecular_name, DESCRIPTION = concise_description, 
                  CORRELATION = abs_spearman_cor) %>%
  dplyr::distinct()
  
  DT::datatable(peak_2,
                filter = "top",
                extensions = "Responsive")
}

```

`r if (eval_cell_3) '------------------------'`

`r if (eval_cell_3) '## Peak 3'`

```{r Peak3, warning=FALSE, message=FALSE}
if(nrow(peaks) > 2){
  peak_3 <- interval_variants %>%  
  dplyr::mutate(query = paste0(CHROM, ":",startPOS, "-",endPOS)) %>% 
  dplyr::filter(query == peaks$interval[3]) %>%
    dplyr::select(CHROM, POS, INTERVAL = query, GENE_NAME = gene_name, REF, ALT, 
                  NT_CHANGE = nt_change, AA_CHANGE = aa_change, GENE_ID = gene_id, 
                  TRANSCRIPT = molecular_name, DESCRIPTION = concise_description, 
                  CORRELATION = abs_spearman_cor) %>%
  dplyr::distinct()
  
  DT::datatable(peak_3,
                filter = "top",
                extensions = "Responsive")
}

```



